dask/array/routines.py

```diff
@@ -160,6 +160,59 @@ def dot(a, b):
     return tensordot(a, b, axes=((a.ndim - 1,), (b.ndim - 2,)))
 
 
+def _func1d_fmt(arr,
+                func1d,
+                func1d_args,
+                func1d_kwargs,
+                ndim_before=0,
+                ndim_after=0):
+    # Drop singleton dimensions from map_blocks to get 1D array.
+    arr1d = arr.flatten()
+
+    # Compute result and ensure it is an array (consistent with NumPy).
+    res = np.asarray(func1d(arr1d, *func1d_args, **func1d_kwargs))
+
+    # Tack on singleton dimensions that map_blocks will expect.
+    res = res[ndim_before * (None,) + (Ellipsis,) + ndim_after * (None,)]
+
+    return res
+
+
+@wraps(np.apply_along_axis)
+def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+    arr = asarray(arr)
+
+    # Validate and normalize axis.
+    arr.shape[axis]
+    axis = len(arr.shape[:axis])
+
+    # Test out some data with the function.
+    test_data = np.ones((1,), dtype=arr.dtype)
+    test_result = np.array(func1d(test_data))
+
+    # Rechunk so that func1d is applied over the full axis.
+    arr = arr.rechunk(
+        axis * (1,) + arr.shape[axis:axis + 1] + (arr.ndim - axis) * (1,)
+    )
```

Comment:
I'm confused about the `(1,)` chunking at the beginning and end here.  

There is also the case to think about where we are drastically increasing average chunk size.  It might be worth reducing some of the other chunkings if the difference is drastic.

Reply:
This also relates to https://github.com/dask/dask/issues/2689

Reply:
There is not much we can do about this as `func1d` is a function that acts on a 1-D array the length of `arr.shape[axis]`. So this chunking is necessary to ensure `func1d` gets arrays of that size and dimensionality. As long as we plan to use `map_blocks` to do implement `apply_along_axis`, we are going to need this chunking.

Reply:
Ah, perhaps we can wrap this in another function that would do a second round of broadcasting locally?

Reply:
Yeah was going to mention that in the previous comment, but guess I didn't. It certainly is an option. Just worry a bit about getting too many "smarts" baked into this function and am unclear on the performance trade-offs ATM. Though I could be worrying too much. 😅

Reply:
So I have made some changes such that only `axis` should be rechunked. The rest should be identities. Managed to cutdown some cruft in the process.

There is still a concern about whether this will performant generally. Though I don't know any good solution beyond passing this decision on to the user. They can always rechunk their Dask Array before passing it into `apply_along_axis` as needed.